src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,433 @@
+            // Validate deadline before creating any objects that require cleanup
+            ValidateDeadline(options.Deadline);
+
+            _callCts = new CancellationTokenSource();
+            Method = method;
+            Options = options;
+            _clock = clock;
+
+            if (options.CancellationToken.CanBeCanceled)
+            {
+                // The cancellation token will cancel the call CTS
+                _ctsRegistration = options.CancellationToken.Register(CancelCall);
+            }
+
+            if (options.Deadline != null && options.Deadline != DateTime.MaxValue)
+            {
+                var timeout = options.Deadline.Value - _clock.UtcNow;
+                _timeout = (timeout > TimeSpan.Zero) ? timeout : TimeSpan.Zero;
+            }
+
+            if (_timeout != null)
+            {
+                // Deadline timer will cancel the call CTS
+                _deadlineTimer = new Timer(DeadlineExceeded, null, _timeout.Value, Timeout.InfiniteTimeSpan);
+            }
+        }
+
+        private void ValidateDeadline(DateTime? deadline)
+        {
+            if (deadline != null && deadline != DateTime.MaxValue && deadline != DateTime.MinValue && deadline.Value.Kind != DateTimeKind.Utc)
+            {
+                throw new InvalidOperationException("Deadline must have a kind DateTimeKind.Utc or be equal to DateTime.MaxValue or DateTime.MinValue.");
+            }
+        }
+
+        public CancellationToken CancellationToken
+        {
+            get { return _callCts.Token; }
+        }
+
+        public bool IsCancellationRequested
+        {
+            get { return _callCts.IsCancellationRequested; }
+        }
+
+        public void StartUnary(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+        }
+
+        public void StartClientStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+        }
+
+        public void StartServerStreaming(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void StartDuplexStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void Dispose()
+        {
+            if (!Disposed)
```

Comment:
It's not super clear to me what the threading model is for this (since you've got cancellation and timeouts). How does this work without locking?

Reply:
I'm concerned about this too.

As long as the GrpcCall object itself is accessed from a single logical thread and then disposed after it's used, `_ctsRegistration?.Dispose()` should block until any possibly-already-triggered calls to `CancelCall` complete. This effectively locks on our behalf.

I don't know if `_deadlineTimer?.Dispose()` will similarly wait for `DeadlineExceeded` to complete. I don't think it does. It would be tempting to use a lock in Dispose and DeadlineExceeded to protect against ODE's, but you risk deadlocking if you're not extremely careful.

After a lot of effort trying to properly dispose the CTS backing the RequestAborted token without ODEs or deadlocking, @jkotalik and I found the best way to avoid deadlocking is to simply not even try disposing CTS that's already been canceled. https://github.com/aspnet/AspNetCore/pull/9333

Reply:
I don't think you need to do anything as complicated as what we did for the RequestAborted token since you don't try to lazily allocate the CTS nor do you expose the CancellationToken to the end user.

I do think that using a CTS instead of a Timer for DeadlineExceeded might help you avoid an ODE in DeadlineExceeded though.

Reply:
Dispose, cancellation and error handling is immature at the moment.

Dispose can happen 3 ways:
1. Response headers are invalid (bad HttpStatusCode or bad content type)
2. Response is received and we finish reading the body
3. User explicitly calls dispose on the public API, which gets routed here

Then there are additional ways _callCts can be canceled:
1. The deadline could be exceeded
2. The user's passed in cancellation token is cancelled